pkg/planner/core/rule_outer_to_inner_join.go

```diff
@@ -0,0 +1,196 @@
+	p := s.self
+	for i, child := range p.Children() {
+		newChild := child.ConvertOuterToInnerJoin(predicates)
+		p.SetChild(i, newChild)
+	}
+	return p
+}
+
+// ConvertOuterToInnerJoin implements base.LogicalPlan ConvertOuterToInnerJoin interface.
+func (p *LogicalJoin) ConvertOuterToInnerJoin(predicates []expression.Expression) base.LogicalPlan {
+	innerTable := p.Children()[0]
+	outerTable := p.Children()[1]
+	switchChild := false
+
+	if p.JoinType == LeftOuterJoin {
+		innerTable, outerTable = outerTable, innerTable
+		switchChild = true
+	}
+
+	// First, simplify this join
+	if p.JoinType == LeftOuterJoin || p.JoinType == RightOuterJoin {
+		canBeSimplified := false
+		for _, expr := range predicates {
+			isOk := isNullFiltered(p.SCtx(), innerTable.Schema(), expr, outerTable.Schema())
+			if isOk {
+				canBeSimplified = true
+				break
+			}
+		}
+		if canBeSimplified {
+			p.JoinType = InnerJoin
+		}
+	}
+
+	// Next simplify join children
+
+	combinedCond := mergeOnClausePredicates(p, predicates)
+	if p.JoinType == LeftOuterJoin || p.JoinType == RightOuterJoin {
+		innerTable = innerTable.ConvertOuterToInnerJoin(combinedCond)
+		outerTable = outerTable.ConvertOuterToInnerJoin(predicates)
+	} else if p.JoinType == InnerJoin || p.JoinType == SemiJoin {
+		innerTable = innerTable.ConvertOuterToInnerJoin(combinedCond)
+		outerTable = outerTable.ConvertOuterToInnerJoin(combinedCond)
+	} else if p.JoinType == AntiSemiJoin {
+		innerTable = innerTable.ConvertOuterToInnerJoin(predicates)
+		outerTable = outerTable.ConvertOuterToInnerJoin(combinedCond)
+	} else {
+		innerTable = innerTable.ConvertOuterToInnerJoin(predicates)
+		outerTable = outerTable.ConvertOuterToInnerJoin(predicates)
+	}
+
+	if switchChild {
+		p.SetChild(0, outerTable)
+		p.SetChild(1, innerTable)
+	} else {
+		p.SetChild(0, innerTable)
+		p.SetChild(1, outerTable)
+	}
+
+	return p
+}
+
+func (*convertOuterToInnerJoin) name() string {
+	return "convert_outer_to_inner_joins"
+}
+
+// ConvertOuterToInnerJoin implements base.LogicalPlan ConvertOuterToInnerJoin interface.
+func (s *LogicalSelection) ConvertOuterToInnerJoin(predicates []expression.Expression) base.LogicalPlan {
+	p := s.self.(*LogicalSelection)
+	combinedCond := append(predicates, p.Conditions...)
+	child := p.Children()[0]
+	child = child.ConvertOuterToInnerJoin(combinedCond)
+	p.SetChildren(child)
+	return p
+}
+
+// ConvertOuterToInnerJoin implements base.LogicalPlan ConvertOuterToInnerJoin interface.
+func (s *LogicalProjection) ConvertOuterToInnerJoin(predicates []expression.Expression) base.LogicalPlan {
+	p := s.self.(*LogicalProjection)
+	canBePushed, _ := BreakDownPredicates(p, predicates)
+	child := p.Children()[0]
+	child = child.ConvertOuterToInnerJoin(canBePushed)
+	p.SetChildren(child)
+	return p
+}
+
+// allConstants checks if only the expression has only constants.
+func allConstants(expr expression.Expression) bool {
+	switch v := expr.(type) {
```

Comment:
To prevent potential bug on Plan Cache, better to add this check as well:
```
	if expression.MaybeOverOptimized4PlanCache(ctx, []expression.Expression{expr}) {
		return false // expression contains non-deterministic parameter
	}
```

Reply:
Done

Reply:
> To prevent potential bug on Plan Cache, better to add this check as well:
> 
> ```
> 	if expression.MaybeOverOptimized4PlanCache(ctx, []expression.Expression{expr}) {
> 		return false // expression contains non-deterministic parameter
> 	}
> ```

Done